### PR TITLE
Add a small speed penalty to backpedaling 

### PR DESCRIPTION
**Pull request by EmoGarbage404**
## About the PR

Backpedaling now applies a small (10%) speed penalty.

## Why / Balance

Generally, combat is inundated with "jousting" where players rush at each other, land a single hit, and back off to wait off their cooldowns. This generally makes attack speed a bit irrelevant, since you are easily able to retreat for the delay without putting yourself at a disadvantage or compromising anything.

By applying a penalty when backpedaling, you put yourself at a disadvantage to your attacker, which encourages more interesting combat than simply pretending to be beyblades. 

## Technical details

MoverController now checks if the player velocity is within a certain angle relative to the direction they're facing and, if true, applies the speed penalty. The current default angle is 135*, which covers straight backwards as well as back-left and back-right.

## Media

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

### CODE REVIEW

**Comment by Rainbeon** [comment]
Can't you just simply look in the backwards direction while moving back since turning is also instantaneous? For example, hit, move mouse a bit while waiting for CD and do the same movement things as usual?

**Comment by K-Dynamic** [comment]
The tf2 approach :trollface: 

Tbh I doubt melee jousting would be affected much by 10% backpedal speed debuff

**Comment by asperger-sind** [comment]
>Yes but this is obviously more work and requires more skill to pull off consistently.

The amount of "extra work" required is negligible, you put your mouse into the opposite cardinal direction and back to where it was after every hit, this also falls into muscle memory territory very fast and just makes players beyblade more.

>It's also going to require that you move your mouse off the target and would likely make attacking again more difficult

Not really? Most people use wideswings to combat and that requires you to aim at the cardinal direction of the person you're trying to hit rather than actually click on them.

>Not to mention that you'd be doing all of this within the span of a second or two.

What does it matter if it's already in muscle memory, this just seems to be conjuring an additional field you need to be good at, which is needless as jousting is going to be better than not jousting regardless because of how jousting conceptually works.

You weave in and out of the opponent's melee range according to your weapon cooldowns in order to minimize damage taken and maximize damage dealt in a very effective way. Besides, you can play around jousting assuming you have a better (and faster) weapon by not responding with a joust and going directly for your opponent, if your speed is matched or higher you force them off the jousting strategy because they already weaved into your melee range and you can keep hitting them.

**Comment by Rainbeon** [comment]
> > Can't you just simply look in the backwards direction while moving back since turning is also instantaneous here? For example, hit, move mouse a bit while waiting for CD and do the same movement things as usual?
> 
> yes but this is obviously more work and requires more skill to pull off consistently. It's also going to require that you move your mouse off the target and would likely make attacking again more difficult. Not to mention that you'd be doing all of this within the span of a second or two.

Moving the mouse is really not that much more difficult or skillful (atleast in any meaningful way, it just shows that you have decent mouse control to be able to flick vaguely then vaguely right over and over) and the only real effect I see of this is it just makes meta combat look even more silly by having all the models twirling as if they're dancing on top of the jousting

**Comment by metalgearsloth** [comment]
> makes meta combat look even more silly

People already look silly twirling over the station.

**Comment by LordEclipse** [comment]
> > makes meta combat look even more silly
> 
> People already look silly twirling over the station.

Ballerina based combat!
_Chivalry/Mordhau moment_

**Comment by Rainbeon** [comment]
> People already look silly twirling over the station.

Yes people twirl normally for fun and it looks silly, but I don't see a need to make that into a core skill you need for optimal meta combat when it doesn't really add much mechanically

**Comment by K-Dynamic** [comment]
For as long as there's been melee combat there's been jousting (unless grappels and parries get simulated too)

**Comment by Hmeister-fake** [comment]
i suggest a general 15% slowdown when you start moving that goes away when you have been holding a direction for a second, meaning that if you press W, S, W, S you will have a constant slowdown since you reverse your velocity, but W, WD, SD for example would keep the normal movement since you are holding onto that velocity. TLDR slowdown when you move in a new direction.

**Comment by LankLTE** [comment]
Since you can turn around on a dime it feels like this kind of just makes combat less accessible, you can just fling your mouse to the other side of the screen to counteract it completely leaving people who are less capable of doing that for whatever reason punished.

**Comment by metalgearsloth** [comment]
> i suggest a general 15% slowdown when you start moving that goes away when you have been holding a direction for a second, meaning that if you press W, S, W, S you will have a constant slowdown since you reverse your velocity, but W, WD, SD for example would keep the normal movement since you are holding onto that velocity. TLDR slowdown when you move in a new direction.

We literally have movement accel, it's just PJB didn't like anything that wasn't near-instant rampup.

**Comment by Hmeister-fake** [comment]
> We literally have movement accel, it's just PJB didn't like anything that wasn't near-instant rampup.

please do a PR for it then, so that us git gremlins can discuss it

**Comment by asperger-sind** [comment]
Start of discussion in Discord contributors chat
https://discord.com/channels/310555209753690112/770682801607278632/1199847995643015199

Effective end of discussion in Discord contributors chat
https://discord.com/channels/310555209753690112/770682801607278632/1199852318150098984

I cba to compile points like I did for the space pen revert PR so have at thee dedicated reader

File: Content.Shared/Movement/Systems/SharedMoverController.cs [hunk 1]
```diff
@@ -246,13 +246,34 @@ public override void UpdateAfterSolve(bool prediction, float frameTime)
...
+                var upperDist = Math.Abs(Angle.ShortestDistance(velocityAngle, upperAngle));
+                var lowerDist = Math.Abs(Angle.ShortestDistance(velocityAngle, lowerAngle));
+
+                if (upperDist <= halfAngle && lowerDist <= halfAngle * 2 ||
+                    lowerDist <= halfAngle && upperDist <= halfAngle * 2)
+                {
+                    backwardsModifier = moveSpeedComponent.BackwardsMovementModifier;
+                }
+            }
+
```
Comment (metalgearsloth) [inline Content.Shared/Movement/Systems/SharedMoverController.cs:270]: Isn't linearvelocity parent-local.

**Comment by Partmedia** [comment]
I think this could be a small improvement to making combat a bit more interesting for the reasons stated in the PR description. Some have expressed that it would not be too hard to learn the new meta which means that this probably won't hurt combat. At the very least, I think it looks more natural for characters to be walking the direction that they're facing, and this would mechanically encourage that.

**Comment by EmoGarbage404** [comment]
I removed the modifier by default. Current PR just adds support but no mechanical impact.

**Comment by Emisse** [comment]
DNM until mouse look isnt cooked

**Comment by Partmedia** [comment]
I understand how this check works, but it seems a bit clunky.

I've outlined a slightly less clunky algorithm that you could consider implementing and testing (Sorry! Just mathguysing here!)

As before get the worldRot and velocityAngle, but don't `Opposite()` either one but then normalize them. This results in the following picture where `vel` and `rot` are both unit length:

image

Because $A \cdot B = |A|| |B||cos(\theta)$ and since both `vel` and `rot` are unit-length:

`Dot(vel, rot) = cos(τ)`

Okay, how do we check that τ is actually within the purple slice lines defined by `BackwardsAngle`? Define `σ = 180°–BackwardsAngle/2`. Graphically, $\cos{\tau}$ and $\cos{\sigma}$ lie on the x-axis because the project those unit vectors. So now to check if somebody is backwards, all you have to do is:

```
if cos(τ) < cos(σ): // note that cos(τ) = Dot(vel, rot)
  // backwards
```

Note that by the magic of the $\pi/2$ periodicity of the $\cos$ function this handles the bottom angle correctly.

And now this check is fast, because it consists of a dot product (multiply/add) and a compare against a constant.

Now to sanity check that this algorithm works: in vector calculus you may recall that the dot product is positive when a vector is "in the direction" of another vector, 0 when the vectors are orthogonal, and negative when the vectors are opposite of each other. If we define `BackwardsAngle = 180°`, then σ = 180° – 180°/2 = 90°. cos(90°) = 0 and so this is just the "in the direction of check" in the easiest case.

**Comment by Emisse** [comment]
Closed due to feature freeze May 10th-June 14th. Comment to have it reopen after this.

**Pull request closed without merging**